exoscale/loadbalancer.go

```diff
@@ -109,154 +97,193 @@ func (l *loadBalancer) EnsureLoadBalancer(ctx context.Context, _ string, service
-	if err != nil {
-		return err
-	}
-	if err := lb.UpdateService(ctx, lbDef.Services[0]); err != nil {
-		return err
+	if remainingServices == 0 {
+		infof("deleting NLB %q", nlb.Name)
+		return l.p.client.DeleteNetworkLoadBalancer(ctx, zone, nlb.ID)
 	}
 
 	return nil
 }
 
-// EnsureLoadBalancerDeleted deletes the specified load balancer if it
-// exists, returning nil if the load balancer specified either didn't exist or
-// was successfully deleted.
-// This construction is useful because many cloud providers' load balancers
-// have multiple underlying components, meaning a Get could say that the LB
-// doesn't exist even if some part of it is still laying around.
-// Implementations must treat the *v1.Service parameter as read-only and not modify it.
-// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
-func (l *loadBalancer) EnsureLoadBalancerDeleted(ctx context.Context, _ string, service *v1.Service) error {
+// updateLoadBalancer updates the matching Exoscale NLB instance according to the *v1.Service spec provided.
+func (l *loadBalancer) updateLoadBalancer(ctx context.Context, service *v1.Service) error {
 	zone, err := l.getLoadBalancerZone(service)
 	if err != nil {
 		return err
 	}
 
-	lb, err := l.fetchLoadBalancer(ctx, service, zone)
+	nlbUpdate, err := buildLoadBalancerFromAnnotations(service)
 	if err != nil {
-		if err == errLoadBalancerNotFound {
-			return nil
-		}
-
 		return err
 	}
 
-	// We delete the NLB instance only if there is only one NLB service left (i.e. the one we're asked to delete).
-	if len(lb.Services) == 1 {
-		return l.p.client.DeleteNetworkLoadBalancer(ctx, zone, lb.ID)
+	nlbCurrent, err := l.p.client.GetNetworkLoadBalancer(ctx, zone, nlbUpdate.ID)
+	if err != nil {
+		return err
 	}
 
-	// Otherwise, we only delete the NLB service and keep the NLB instance as it holds other services.
-	lbService, err := l.fetchLoadBalancerService(lb, service)
-	if err != nil {
-		if err == errLoadBalancerServiceNotFound {
-			return nil
+	if isLoadBalancerUpdated(nlbCurrent, nlbUpdate) {
+		infof("updating NLB %q", nlbCurrent.Name)
+		if _, err = l.p.client.UpdateNetworkLoadBalancer(ctx, zone, nlbUpdate); err != nil {
+			return err
+		}
+		debugf("NLB %q updated successfully", nlbCurrent.Name)
+	}
+
+	// Delete the NLB services which port is not present in the updated version.
+	nlbServices := make(map[uint16]*egoscale.NetworkLoadBalancerService)
+next:
+	for _, nlbServiceCurrent := range nlbCurrent.Services {
+		for _, nlbServiceUpdate := range nlbUpdate.Services {
+			// If a service exposing the same port already exists,
+			// flag it for update and save its ID for later reference.
+			if nlbServiceUpdate.Port == nlbServiceCurrent.Port {
+				debugf("service port %d already in use by NLB service %q, marking for update",
+					nlbServiceCurrent.Port,
+					nlbServiceCurrent.Name)
+				nlbServices[nlbServiceCurrent.Port] = nlbServiceCurrent
+				continue next
+			}
```

Comment:
This code could be spot into a function, instead of continuing on the  `next` keyword.

Reply:
@pierre-emmanuelJ can you show me how you'd do it with a function?